### ISSUE TITLE
Wrong `this` in template function passed by parent components

### ISSUE THREAD

**Issue by giovannipiller** [ractivejs/ractive#3031]
## Description:
We're in the process of upgrading our Single Page App from 0.8.14 to 0.9.2.

Calling a template function declared in ParentComponent, from a ChildComponent will result in the function having `this === ChildComponent`.

Code below should help figuring out what we're trying to do.

## Versions affected:
0.9.x

## Platforms affected:
Reproduced on Chrome, Safari (macOS)

## Reproduction:
https://jsfiddle.net/tgevy77v/

```javascript
const ChildComponent = Ractive.extend({
  template: '#template_child',

  // getSomeData is passed from the parent component via template
});

const RootComponent = Ractive.extend({
  template: '#template',

  components: { ChildComponent },

  data() {
    return {
      getSomeData() {
        // issue: this === ChildComponent when getSomeData() is called from ChildComponent
        return this.generateData();
      },
    };
  },

  generateData() {
    return 'John';
  }
});

const r = window.r = new RootComponent();
r.render('#main');
```

HTML:
```html
<script type="text/html" id="template">
	<ChildComponent getSomeData="{{getSomeData}}"/>
<pre>
Open the console, there's an error message.
`this` for getSomeData === ChildComponent rather than RootComponent.
</pre>
</script>


<script type="text/html" id="template_child">
	<h1>Hello {{getSomeData()}}!</h1>
</script>

<div id="main"></div>
```

## Workaround
https://jsfiddle.net/fLrpxuv8/

A possible workaround is turning `getSomeData` into an arrow function.
_However_ it wasn't necessary on 0.8.x and before, so it might be worth investigating.

```javascript
data() {
  return {
    // arrow function solves the issue
    getSomeData: () => {
      return this.generateData();
    },
  };
},
```

**Comment by evs-chris** [comment]
This seems to be at odds with #2552, though I'm not entirely sure which is the more expected behavior. With root data, which is generally what mappings are, you get the instance associated with the current keypath as the context of the function. This case _could_ be a refinement of that one, meaning that if you map a root function, it would use the source instance as the binding point. Any opinions?

**Comment by ghost** [comment]
The way it's written, `this` makes sense as the parent. But there's the "it depends on how it's called" side of JS as well. Hmmm... 🤔

**Comment by evs-chris** [comment]
I think the idea is that functions at root are effectively members of the instance and are therefore called in the scope of their instance. It gets tricky when the function exists at more than one root. The reason previous versions didn't stumble into this is that they were wholesale bound to the instance at `set`, so even mapping and rebinding had no effect.

**Comment by giovannipiller** [comment]
> This case could be a refinement of that one, meaning that if you map a root function, it would use the source instance as the binding point. Any opinions?

In my mind, functions created by an instance, should always be bound to it by default.

Note that you can easily do something like this, from any Child component:
`this.get('myRootFunction').call(this)`

However, you can't run it as easily while bound to its "source" instance, unless you know which one that is.

### Example
you may have a list of components like this:
A, B, C, D, E, F

`C` has a function named 'myFunction' that gets passed down up to `F`.
`F` attempts to run 'myFunction', but wants to do so by having it bound to its source component (which would be `C`).
Now, `F` has no way of knowing who's the original source of `myFunction`, unless we explicitly declare it prior to execution. Which is cumbersome, possibly unreliable and not a practical thing to do, especially from a template.

ex. `this.get('myFunction').call(this.findParent('C'))` – however, if you don't know which component created myFunction, you're out of luck.

---

### Example (kinda similar to my previous, but more real-life)
I have a `<Users>` component.

```javascript
const Users = Ractive.extend({

  // [...]

  getUsers() {
    // returns the list of users. Almost no use from template, many uses from within Users itself
  }

  data() {
    return {
      usernameFor(id) {
        const users = this.getUsers();
        return users.find(user => user.id === id).name;
      }
    }
  }

  // [...]

});
```

A child `<Dropdown>` component, further down the hierarchy, attempts to use `usernameFor()`, with no luck.

In short: I feel like functions returned by `data()` should be bound to the instance where they've been created, and not the instance where they're invoked. Should that be a necessity, a simple `.call(this)` would do the trick.

I'm of course biased towards this behaviour, but I can't think of an advantage in doing it the 0.9.2 way. **Does anybody have an example or reasoning to keep it as it is?**

_(strong just for visibility 😊)_
